File: frontend/src/app/_models/segment.ts

export class Segment {
  id: number;
  name: string;
  type: string;
  creationDate: string;
  conversionEvent: string;
  didEvents: DidEvents;
  didNotEvents: DidEvents;
}

export class DidEvents {
  description: string;
  joinCondition: JoinCondition;
  events: Event[];
}

export class JoinCondition {
  anyOfCount: number;
  conditionType: string; // AllOf / AnyOf
}

export class Event {
  name: string;
  dateFilter: DateFilter;
  propertyFilter: PropertyFilter;
  whereFilter: WhereFilter;
}

export class DateFilter {
  operator: DateOperator;
  values: string[];
  valueUnit: string;
}

export class PropertyFilter {
  name: string;
  type: PropertyType;
  filterType: PropertyFilterType;
  operator: string;
  values: string[];
  valueUnit: string;
}

export enum PropertyType {
  string="string",
  number="number",
  date="date"
}

export enum PropertyFilterType {
  eventproperty="eventproperty",
  genericproperty="genericproperty",
  UTM="UTM"
}

export class WhereFilter {
  count: WhereCount;
  sumOfValuesOf: PropertySumOf;
}

export class WhereCount {
  operator: NumberOperator;
  values: number[]
}

export class PropertySumOf {
  propertyName: string;
  operator: NumberOperator;
  values: number[]
}

export enum DateOperator {
  Before="Before",
  After="After",
  On="On",
  Between="Between",
  InThePast="InThePast",
  WasExactly="WasExactly",
  Today="Today",
  InTheFuture="InTheFuture",
  WillBeExactly="WillBeExactly"
}

export enum NumberOperator {
  Equals="Equals",
  Between="Between",
  GreaterThan="GreaterThan",
  LessThan="LessThan",
  NotEquals="NotEquals",
  Exists="Exists",
  DoesNotExist="DoesNotExist"
}

export enum StringOperator {
  Equals="Equals",
  NotEquals="NotEquals",
  Contains="Contains",
  DoesNotContain="DoesNotContain",
  Exists="Exists",
  DoesNotExist="DoesNotExist"
}

export class GlobalFilters {
  userProperties: UserProperty[];
  demograpgics: Demography[];
  geography: Geography[];
  technographics: Technography[];
  rechability: Reachability;
  appFields: AppFields;
}

export class UserProperty {
  propertyName: string;
  propertyType: string;
  operator: any;
  values: any[];
  valueUnit: string;
}

export class Demography {
  age: string;
  gender: string;
}

export class Geography {
  country: Country;
  state: State;
  city: City;
}

export class Country {
  id: number;
  name: string;
}

export class State {
  id: number;
  name: string;
}

export class City {
  id: number;
  name: string;
}

export class Technography {
  browser: Browser[];
  device: Device[];
  OS: OS[];
}

export enum Browser {
  Chrome = "Chrome",
  Firefox = "Firefox",
  InternetExplorer = "Internet Explorer",
  MobileApplication = "Mobile Application",
  Opera = "Opera",
  Others = "Others",
  Safari = "Safari"
}

export enum Device {
  Desktop = "Desktop",
  Mobile = "Mobile",
  Tablet = "Tablet"
}

export enum OS {
  Android = "Android",
  iOS = "iOS",
  Linux = "Linux",
  MACOS = "MACOS",
  Windows = "Windows",
  Others = "Others"
}

export class Reachability {
  HasDeviceToken: boolean;
  HasEmailAddress: boolean;
  HasPhoneNumber: boolean;
  UnsubscribedPush: boolean;
  UnsubscribedEmail: boolean;
  UnsubscribedSMS: boolean;
}

export class AppFields {
  OSVersion: AppFieldOperator;
  AppVersion: AppFieldOperator;
  Make: AppFieldOperator;
  Model: AppFieldOperator;
  SDKVersion: AppFieldOperator;
}

export class AppFieldOperator {
}

export class RegisteredEvent {
  name: string;
  properties: RegisteredEventProperties[];
}

export class RegisteredEventProperties {
  name: string;
  dataType: string;
  regex: string;
  options: any[];
}

export class GlobalFilterItem {
  value: string;
  displayName: string;
  type: string;
}

File: src/main/kotlin/com/und/web/model/Segment.kt
package com.und.web.model

import java.time.LocalDateTime


class Segment {

    var id: Long? = null

    var name: String = ""
    var type: String = ""
    var creationDate: LocalDateTime = LocalDateTime.now()
    var conversionEvent: String? = null
    var didEvents: DidEvents? = null
    var didNotEvents: DidEvents? = null
    var globalFilters: List<GlobalFilter> = listOf()
    var geographyFilters: List<Geography> = listOf()
}

class DidEvents {
    var description: String? = null
    var joinCondition: JoinCondition = JoinCondition()
    var events: List<Event> = listOf()
}

class JoinCondition {
    var anyOfCount: Long? = null
    var conditionType: ConditionType = ConditionType.AllOf // AnyOf
}

enum class ConditionType {
    AllOf,
    AnyOf
}

class Event {
    var name: String = ""
    var dateFilter: DateFilter = DateFilter()
    var propertyFilters: List<PropertyFilter> = listOf()
    var whereFilter: WhereFilter? = null
}

class DateFilter {
    var operator: DateOperator = DateOperator.After
    var values: List<String> = listOf()
    var valueUnit: String = ""
}

class PropertyFilter {
    var name: String = ""
    var type: DataType = DataType.string
    private var _filterType: PropertyFilterType? = null
    var filterType: PropertyFilterType? = null
        get() {
            if (_filterType == null) {
                val genericProperties = listOf("First Time", "Time of day",
                        "Day of the week",
                        "Day of the month")
                val UTMProperties = listOf("UTM Source",
                        "UTM Visited")
                _filterType = when{
                    genericProperties.contains(name) -> PropertyFilterType.genericproperty
                    UTMProperties.contains(name) -> PropertyFilterType.UTM
                    else ->  PropertyFilterType.eventproperty
                }

            }
            return _filterType
        }

    var operator: String = ""
    var values: List<String> = listOf()
    var valueUnit: String = ""
}


enum class PropertyFilterType {
    eventproperty,
    genericproperty,
    UTM
}

class WhereFilter {
    var whereFilterName: WhereFilterName? = null
    var propertyName: String = ""
    var operator: NumberOperator? = null
    var values: List<Long>? = null
}

enum class WhereFilterName {
    Count,
    SumOfValuesOf
}

enum class DateOperator {
    Before,
    After,
    On,
    Between,
    InThePast,
    WasExactly,
    Today,
    InTheFuture,
    WillBeExactly
}

enum class NumberOperator {
    Equals,
    Between,
    GreaterThan,
    LessThan,
    NotEquals,
    Exists,
    DoesNotExist
}

enum class StringOperator {
    Equals,
    NotEquals,
    Contains,
    DoesNotContain,
    Exists,
    DoesNotExist
}

class GlobalFilter {
    var globalFilterType: GlobalFilterType = GlobalFilterType.AppFields
    var name: String = ""
    var type: String = ""
    var operator: String = ""
    var values: Array<Any> = arrayOf()
    var valueUnit: String = ""
}

enum class GlobalFilterType(val type: String) {
    UserProperties("UserProperties"),
    Demographics("Demographics"),
    Technographics("Technographics"),
    Reachability("Reachability"),
    AppFields("appFields")
}

class Geography {
    var country: Country = Country()
    var state: State = State()
    var city: City = City()
}

class Country {
    var id: Long = 0
    var name: String = ""
}

class State {
    var id: Long = 0
    var name: String = ""
}

class City {
    var id: Long = 0
    var name: String = ""
}

class RegisteredEvent {
    var name: String = ""
    var properties: List<RegisteredEventProperties> = listOf()
}

class RegisteredEventProperties {
    var name: String = ""
    var dataType:DataType  = DataType.string
    var regex: String = ""
    var options: Array<Any> = arrayOf()
}

class GlobalFilterItem {
    var value: String = ""
    var displayName: String = ""
    var type: String = ""
}


enum class DataType{
    string,
    number,
    date,
    range,
    boolean
}